core/test/ParaGrid_test.cpp

```diff
@@ -0,0 +1,321 @@
+    REQUIRE(coordinates.components({ 12, 13 })[1] - coordinates.components({ 12, 12 })[1] == scale);
+
+    ModelState state = { {
+                             { maskName, mask },
+                             { hiceName, hice },
+                             { ciceName, cice },
+                             { hsnowName, hsnow },
+                             { ticeName, tice },
+                             { coordsName, coordinates },
+                         },
+        {} };
+
+    ModelMetadata metadata;
+    metadata.setTime(TimePoint("2000-01-01T00:00:00Z"));
+
+    grid.dumpModelState(state, metadata, filename, true);
+
+    REQUIRE(std::filesystem::exists(std::filesystem::path(filename)));
+
+    // Reset the array dimensions to make sure that the read function gets them correct
+    ModelArray::setDimension(ModelArray::Dimension::X, 1);
+    ModelArray::setDimension(ModelArray::Dimension::Y, 1);
+    ModelArray::setDimension(ModelArray::Dimension::Z, 1);
+    ModelArray::setDimension(ModelArray::Dimension::XVERTEX, 1);
+    ModelArray::setDimension(ModelArray::Dimension::YVERTEX, 1);
+    ModelArray::setDimension(ModelArray::Dimension::XCG, 1);
+    ModelArray::setDimension(ModelArray::Dimension::YCG, 1);
+    // In the full model numbers of DG components are set at compile time, so they are not reset
+    REQUIRE(ModelArray::nComponents(ModelArray::Type::DG) == DG);
+    REQUIRE(ModelArray::nComponents(ModelArray::Type::VERTEX) == ModelArray::nCoords);
+
+    ParametricGrid gridIn;
+    ParaGridIO* readIO = new ParaGridIO(gridIn);
+    gridIn.setIO(readIO);
+
+    ModelState ms = gridIn.getModelState(filename);
+
+    REQUIRE(ModelArray::dimensions(ModelArray::Type::Z)[0] == nx);
+    REQUIRE(ModelArray::dimensions(ModelArray::Type::Z)[1] == ny);
+    REQUIRE(ModelArray::dimensions(ModelArray::Type::Z)[2] == nz);
+
+    REQUIRE(ms.data.size() == state.data.size());
+
+    ModelArray& ticeRef = ms.data.at(ticeName);
+    REQUIRE(ModelArray::nDimensions(ModelArray::Type::Z) == 3);
+    REQUIRE(ticeRef.getType() == ModelArray::Type::Z);
+    REQUIRE(ticeRef.nDimensions() == 3);
+    REQUIRE(ticeRef.dimensions()[0] == nx);
+    REQUIRE(ticeRef.dimensions()[1] == ny);
+    REQUIRE(ticeRef.dimensions()[2] == nz);
+
+    ModelArray& hiceRef = ms.data.at(hiceName);
+    REQUIRE(hiceRef.nDimensions() == 2);
+    REQUIRE(hiceRef.dimensions()[0] == nx);
+    REQUIRE(hiceRef.dimensions()[1] == ny);
+    REQUIRE(ModelArray::nComponents(ModelArray::Type::DG) == DG);
+    REQUIRE(hiceRef.nComponents() == DG);
+
+    REQUIRE(ticeRef(12, 14, 1) == tice(12, 14, 1));
+
+    ModelArray& coordRef = ms.data.at(coordsName);
+    REQUIRE(coordRef.nDimensions() == 2);
+    REQUIRE(coordRef.nComponents() == 2);
+    REQUIRE(coordRef.components({ 12, 13 })[0] - coordRef.components({ 11, 13 })[0] == scale);
+    REQUIRE(coordRef.components({ 12, 13 })[1] - coordRef.components({ 12, 12 })[1] == scale);
+
+    std::filesystem::remove(filename);
```

Comment:
Comment: Perhaps worth doing this teardown as part of catch2 destructor (using sections and writing this below all sections). In event of an error we're left with junk not deleted. If you've got time I'd do it here but could be worth assessing and re-jigging some tests as a separate issue. Might also remove some code duplication by sharing setup (although my inspection here hasn't been that thorough, my guess is it would be some work to get from here to a shared test setup in this file).

Reply:
Yes, I need to review Catch2 setup and teardown, and actually use those facilities.

Reply:
From what I can tell it's a pain and you don't get much for free. I've tried in #207 to do setup and teardown in test_section and have the tests running inside sections each time doing a clean setup and tear down (and looping over sections through the dynamic section, but ignore that for the purpose of this comment). Can't seem to get much better than that though. If you find something that takes a snapshot of the environment and restores it automatically, let me know..... (otherwise maybe we might want to make one --- and undo what I did in 207).

Reply:
Realistically, how much of an issue is this? If the tests fail in the test environment, then the file disappears when the environment does. On a dev's machine the file will finally be deleted when they fix the test and get it running to completion. It's an incentive!

Reply:
Depends on the user and how big the project gets but this is really about future-proofing. For us, in CI or dev, the main impact it could have is on subsequent tests --- right now, I don't think we have many tests which would be impacted by this but if we need to validate the existence and content of output files then we start running the risk that we are actually assessing the output of previous tests rather than the test we are running. If two tests produce files and they were to share a name (or if we wanted to swap this test to dynamic sections or just split into sections) which may not be an issue here but as the project gets bigger we lose control over that / have to begin to check that for each test we write. Similarly if part 1 of a test succeeds and part 2 fails and we make changes to source to try and fix it, we could find that we've stopped producing the file or create it in a different location by accident but be under the impression our change has had no effect, while a new failure in part 1 would make it very very obvious we have broken something. In CI this is less of a big deal but obvs our dev environment persists.

For the end-user, they'll probably want to run tests to validate on their machine before doing any simulations. In such a case, it would be easier not to clutter folders with test artefacts, particularly if end-users can add functionality (and tests). If someone (including us) were to rewrite the github action workflow and run tests from the top level directory (which is attractive to simplify the file, but not necessarily what we'd want to do), this would become more critical. In general though, this test can be run from any directory and that means a user could and somewhere probably would end up doing that. Additionally in any of these scenarios, if the test later depended on another file (perhaps just to compare numerical results), it could fail as the path to file might be wrong and we'd have left the file behind.

And for any future development, the requirements/limitations of any feature are really captured (even enforced) by tests. If you wanted to see how a new feature might impact other features, you'd probably want to make a dummy change and run some tests and build a list of components/methods impacted which you'd need to modify. This would definitely be the case if we were removing support for anything - you would intentionally be triggering failures and don't want any accidental passes as a result of that. You want to be able to 100% trust that the test results reflect interdependencies in the code and that the test ecosystem has no impact. 

Dream scenario is to make a temporary dir at the start of each test, copy inputs/configs (if applicable) into that folder and run the tests there. On success or failure, we'd then remove the dir and contents and return to original directory. Obvs this is a task for later because it's not a necessary thing but it makes life easier as the project evolves (and, like all things, is easier to do earlier than later)

Before doing anything it'd be worth checking if Catch2 already does this. I couldn't see it but it would make the most sense if there was an environment setup macro (or something) it ran before re-doing the setup (top of TEST_CASE) before running the next SECTION (and then a cleanup macro after the end of the test case after exiting the section)